File: EPI/14.3/swift/14.3_test.swift
import Darwin

func == <T: Equatable, U: Equatable>(lhs: [(T, U)], rhs: [(T, U)]) -> Bool {
    guard lhs.count == rhs.count else {
        return false
    }
    for (i, lt) in lhs.enumerated() {
        let rt = rhs[i]
        guard lt.0 == rt.0 && lt.1 == rt.1 else {
            return false
        }
    }
    return true
}

let tests: [(string: String, occurrences: [Occurrences])] = [
    (
        string: "",
        occurrences: []
    ),
    (
        string: "a",
        occurrences: [("a", 1)]
    ),
    (
        string: "aa",
        occurrences: [("a", 2)]
    ),
    (
        string: "ab",
        occurrences: [("a", 1), ("b", 1)]
    ),
    (
        string: "ba",
        occurrences: [("a", 1), ("b", 1)]
    ),
    (
        string: "bcdacebe",
        occurrences: [("a", 1), ("b", 2), ("c", 2), ("d", 1), ("e", 2)]
    ),
]

for test in tests {
    let actual = occurrences(test.string)
    guard actual == test.occurrences else {
        print("For test string \(test.string), expected occurrences to be \(test.occurrences), " +
            "but were \(actual)")
        exit(1)
    }
}

File: EPI/14.3/typescript/14.3_test.ts
import occurrences, {Occurrence} from './14.3';

declare function require(name: String): any;
const assert = require('assert');
const inspect = (v: any) => require('util').inspect(v, {depth: null});

type Test = {string: string, occurrences: Occurrence[]};

const tests: Test[] = [
    {
        string: '',
        occurrences: []
    },
    {
        string: 'a',
        occurrences: [
            {char: 'a', count: 1},
        ]
    },
    {
        string: 'aa',
        occurrences: [
            {char: 'a', count: 2},
        ]
    },
    {
        string: 'ab',
        occurrences: [
            {char: 'a', count: 1},
            {char: 'b', count: 1},
        ]
    },
    {
        string: 'ba',
        occurrences: [
            {char: 'a', count: 1},
            {char: 'b', count: 1},
        ]
    },
    {
        string: 'bcdacebe',
        occurrences: [
            {char: 'a', count: 1},
            {char: 'b', count: 2},
            {char: 'c', count: 2},
            {char: 'd', count: 1},
            {char: 'e', count: 2},
        ]
    },
];

for (const test of tests) {
    const actual = occurrences(test.string);
    assert.deepStrictEqual(actual, test.occurrences, `For test string ${test.string}, ` +
        `expected occurrences to be ${inspect(test.occurrences)}, but were ${inspect(actual)}`);
}
